File: packages/core/dod/dod.test.ts
import { describe, expect, it } from "vitest";
import { failedAssertions, milestoneSatisfied } from "./index";

const a = (id: string) => ({ id });
const r = (assertionId: string, passed: boolean) => ({ assertionId, passed });

describe("milestoneSatisfied", () => {
  it("is vacuously satisfied with no assertions", () => {
    expect(milestoneSatisfied([], [])).toBe(true);
  });

  it("is satisfied when every assertion has a passing verdict", () => {
    expect(milestoneSatisfied([a("x"), a("y")], [r("x", true), r("y", true)])).toBe(true);
  });

  it("is unsatisfied when an assertion has no verdict", () => {
    expect(milestoneSatisfied([a("x"), a("y")], [r("x", true)])).toBe(false);
  });

  it("is unsatisfied when an assertion fails", () => {
    expect(milestoneSatisfied([a("x")], [r("x", false)])).toBe(false);
  });

  it("treats any failing verdict for an assertion as a failure", () => {
    expect(milestoneSatisfied([a("x")], [r("x", true), r("x", false)])).toBe(false);
  });

  it("ignores stray verdicts for unknown assertions", () => {
    expect(milestoneSatisfied([a("x")], [r("x", true), r("ghost", false)])).toBe(true);
  });
});

describe("failedAssertions", () => {
  it("returns the unsatisfied assertions in input order", () => {
    expect(failedAssertions([a("x"), a("y"), a("z")], [r("y", true)])).toEqual([a("x"), a("z")]);
  });

  it("returns empty when all pass", () => {
    expect(failedAssertions([a("x")], [r("x", true)])).toEqual([]);
  });
});

File: server/internal/dod/dod_test.go
package dod

import (
	"reflect"
	"testing"
)

func a(id string) Assertion       { return Assertion{ID: id} }
func r(id string, ok bool) Result { return Result{AssertionID: id, Passed: ok} }

func TestMilestoneSatisfied(t *testing.T) {
	cases := []struct {
		name       string
		assertions []Assertion
		results    []Result
		want       bool
	}{
		{"no assertions is vacuously satisfied", nil, nil, true},
		{"single passing assertion", []Assertion{a("x")}, []Result{r("x", true)}, true},
		{"single unvalidated assertion fails", []Assertion{a("x")}, nil, false},
		{"single failing assertion fails", []Assertion{a("x")}, []Result{r("x", false)}, false},
		{
			"all passing",
			[]Assertion{a("x"), a("y")},
			[]Result{r("x", true), r("y", true)},
			true,
		},
		{
			"one of two failing",
			[]Assertion{a("x"), a("y")},
			[]Result{r("x", true), r("y", false)},
			false,
		},
		{
			"missing one verdict fails",
			[]Assertion{a("x"), a("y")},
			[]Result{r("x", true)},
			false,
		},
		{
			"any failing verdict for an assertion fails it",
			[]Assertion{a("x")},
			[]Result{r("x", true), r("x", false)},
			false,
		},
		{
			"stray result for unknown assertion is ignored",
			[]Assertion{a("x")},
			[]Result{r("x", true), r("ghost", false)},
			true,
		},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if got := MilestoneSatisfied(c.assertions, c.results); got != c.want {
				t.Errorf("MilestoneSatisfied() = %v, want %v", got, c.want)
			}
		})
	}
}

func TestFailedAssertions(t *testing.T) {
	cases := []struct {
		name       string
		assertions []Assertion
		results    []Result
		want       []string
	}{
		{"none failed when all pass", []Assertion{a("x")}, []Result{r("x", true)}, nil},
		{"unvalidated assertion is failed", []Assertion{a("x")}, nil, []string{"x"}},
		{
			"preserves input order of failures",
			[]Assertion{a("x"), a("y"), a("z")},
			[]Result{r("y", true)},
			[]string{"x", "z"},
		},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			got := ids(FailedAssertions(c.assertions, c.results))
			if !reflect.DeepEqual(got, c.want) {
				t.Errorf("FailedAssertions() = %v, want %v", got, c.want)
			}
		})
	}
}

func ids(as []Assertion) []string {
	if len(as) == 0 {
		return nil
	}
	out := make([]string, len(as))
	for i, x := range as {
		out[i] = x.ID
	}
	return out
}
